File: 19 Favorite Code Examples/FeatureSelection/SolidWorksMacro.vb
Imports SolidWorks.Interop.sldworks
Imports SolidWorks.Interop.swconst
Imports System

Partial Class SolidWorksMacro

  Public Sub main()
    Dim Model As ModelDoc2 = swApp.ActiveDoc
    'make sure the active document is a part
    If Model Is Nothing Then
      MsgBox("Please open a part first.", MsgBoxStyle.Exclamation)
      Exit Sub
    ElseIf Model.GetType <> swDocumentTypes_e.swDocPART Then
      MsgBox("For parts only.", MsgBoxStyle.Exclamation)
      Exit Sub
    End If
    Dim Part As PartDoc = Model
    'get the Extension interface from the model
    Dim Ext As ModelDocExtension = Model.Extension

    'get the first feature in the FeatureManager
    Dim feat As Feature = Part.FeatureByName("Extrude1")
    If Not feat Is Nothing Then
      'select it
      feat.Select2(False, -1)
      'get persistent ID
      Dim IDCount As Integer = Ext.GetPersistReferenceCount3(feat)
      Dim ID(IDCount) As Byte
      ID = Ext.GetPersistReference3(feat)
      Stop 'check selected feature and ID

      'get rid of the feature reference
      feat = Nothing
      'clear the selection
      Model.ClearSelection2(True)
      'select the feature by its ID
      Dim errors As Integer
      'get the feature by its persistent ID
      feat = Ext.GetObjectByPersistReference3(ID, errors)
      feat.Select2(False, -1)
      Stop 'check selected feature
    End If
  End Sub


  ''' <summary>
  ''' The SldWorks swApp variable is pre-assigned for you.
  ''' </summary>
  Public swApp As SldWorks


End Class

File: 19 Favorite Code Examples/FeatureSelection/SolidWorksMacro.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Windows;
using System.Windows.Forms;
using System.Diagnostics;
using SolidWorks.Interop.sldworks;
using SolidWorks.Interop.swconst;


namespace FeatureSelection
{
    public partial class SolidWorksMacro
    {
        public void Main()
        {
            ModelDoc2 Model = (ModelDoc2)swApp.ActiveDoc;
            // make sure the active document is a part
            if (Model == null)
            {
                swApp.SendMsgToUser2("Please open a part first.",
                    (int)swMessageBoxIcon_e.swMbWarning,
                    (int)swMessageBoxBtn_e.swMbOk);
                return;
            }
            else if (Model.GetType() != (int)swDocumentTypes_e.swDocPART)
            {
                swApp.SendMsgToUser2("For parts only.",
                    (int)swMessageBoxIcon_e.swMbWarning,
                    (int)swMessageBoxBtn_e.swMbOk);
                return;
            }

            PartDoc Part = (PartDoc)Model;
            // get the Extension interface from the model
            ModelDocExtension Ext = Model.Extension;

            // get a feature in the FeatureManager
            // assuming a feature named "Extrude1"
            Feature feat = (Feature)Part.FeatureByName("Extrude1");
            if (feat != null)
            {
                // select it
                feat.Select2(false, -1);
                // get persistent ID
                int IDCount = Ext.GetPersistReferenceCount3(feat);
                object ID = null;
                ID = Ext.GetPersistReference3(feat);
                Debugger.Break(); // check selected feature and ID

                // get rid of the feature reference
                feat = null;
                // clear the selection
                Model.ClearSelection2(true);
                // select the feature by its ID
                int errors = 0;
                // get the feature by its persistent ID
                feat = (Feature)Ext.GetObjectByPersistReference3(ID, out errors);
                feat.Select2(false, -1);
                Debugger.Break(); // check selected feature
            }
        }


        // The SldWorks swApp variable is pre-assigned for you.
        public SldWorks swApp;

    }
}

